File: python/zeroipc/array.py
"""
Fixed-size array in shared memory with duck typing support.
"""

import struct
import numpy as np
from typing import Optional, Union
from .memory import Memory


class Array:
    """
    Fixed-size array in shared memory.
    
    This implementation uses duck typing - users specify the dtype
    and are responsible for type consistency.
    """
    
    HEADER_FORMAT = '<Q'  # uint64_t capacity
    HEADER_SIZE = struct.calcsize(HEADER_FORMAT)
    
    def __init__(self, memory: Memory, name: str, 
                 capacity: Optional[int] = None, 
                 dtype: Optional[Union[np.dtype, str]] = None):
        """
        Create or open an array.
        
        Args:
            memory: Shared memory instance
            name: Name of the array
            capacity: Number of elements (required for new arrays)
            dtype: NumPy dtype (required)
        
        Raises:
            ValueError: If name is too long or required parameters missing
            TypeError: If dtype is not specified
        """
        if len(name) > 31:
            raise ValueError("Name too long (max 31 characters)")
        
        if dtype is None:
            raise TypeError("dtype is required")
        
        self.memory = memory
        self.name = name
        self.dtype = np.dtype(dtype)
        
        entry = memory.table.find(name)
        
        if entry:
            # Open existing array
            self.offset = entry.offset
            
            # Read capacity from header
            capacity_bytes = memory.mmap[self.offset:self.offset + self.HEADER_SIZE]
            stored_capacity = struct.unpack(self.HEADER_FORMAT, capacity_bytes)[0]
            
            # Validate capacity if provided
            if capacity is not None and capacity != stored_capacity:
                raise ValueError(
                    f"Capacity mismatch: array has {stored_capacity} "
                    f"but requested {capacity}"
                )
            
            self.capacity = stored_capacity
            
        else:
            # Create new array
            if capacity is None:
                raise ValueError("Capacity required to create new array")
            
            self.capacity = capacity
            
            # Allocate space
            total_size = self.HEADER_SIZE + capacity * self.dtype.itemsize
            self.offset = memory.table.allocate(total_size)
            
            # Add to table
            if not memory.table.add(name, self.offset, total_size):
                raise RuntimeError("Failed to add array to table (table full?)")
            
            # Write header
            struct.pack_into(
                self.HEADER_FORMAT, memory.mmap, self.offset, capacity
            )
            
            # Zero-initialize data
            data_offset = self.offset + self.HEADER_SIZE
            data_size = capacity * self.dtype.itemsize
            memory.mmap[data_offset:data_offset + data_size] = bytes(data_size)
        
        # Create NumPy array view of the data
        data_offset = self.offset + self.HEADER_SIZE
        self.array = np.frombuffer(
            memory.mmap,
            dtype=self.dtype,
            count=self.capacity,
            offset=data_offset
        )
    
    def __getitem__(self, index):
        """Get item by index"""
        return self.array[index]
    
    def __setitem__(self, index, value):
        """Set item by index"""
        self.array[index] = value
    
    def __len__(self):
        """Get array length"""
        return self.capacity
    
    def __iter__(self):
        """Iterate over array"""
        return iter(self.array)
    
    def fill(self, value):
        """Fill array with value"""
        self.array[:] = value
    
    @property
    def data(self):
        """Get underlying NumPy array"""
        return self.array
    
    def __repr__(self):
        return f"Array(name='{self.name}', capacity={self.capacity}, dtype={self.dtype})"

File: go/zeroipc/array.go
package zeroipc

import (
	"encoding/binary"
	"errors"
	"fmt"
	"unsafe"
)

// ArrayHeaderSize is the size of the array header in bytes.
const ArrayHeaderSize = 8

// Array is a fixed-size array in shared memory.
// The element type must be a fixed-size type (numeric types).
//
// Binary layout:
//   - capacity: uint64 (8 bytes)
//   - data: capacity * sizeof(T) bytes
type Array[T Numeric] struct {
	memory   *Memory
	name     string
	offset   int
	capacity int
	elemSize int
}

// Numeric constrains Array elements to numeric types that are binary-compatible.
type Numeric interface {
	~int8 | ~int16 | ~int32 | ~int64 |
		~uint8 | ~uint16 | ~uint32 | ~uint64 |
		~float32 | ~float64
}

// NewArray creates a new array in shared memory.
func NewArray[T Numeric](memory *Memory, name string, capacity int) (*Array[T], error) {
	if len(name) >= NameSize {
		return nil, errors.New("name too long (max 31 characters)")
	}

	if capacity <= 0 {
		return nil, errors.New("capacity must be positive")
	}

	var zero T
	elemSize := int(unsafe.Sizeof(zero))

	// Check if already exists
	if entry := memory.Find(name); entry != nil {
		return nil, fmt.Errorf("array '%s' already exists", name)
	}

	// Allocate space
	totalSize := ArrayHeaderSize + capacity*elemSize
	offset, err := memory.Allocate(name, totalSize)
	if err != nil {
		return nil, fmt.Errorf("failed to allocate: %w", err)
	}

	// Write capacity header
	binary.LittleEndian.PutUint64(memory.Data()[offset:offset+8], uint64(capacity))

	// Zero-initialize data
	dataStart := offset + ArrayHeaderSize
	dataEnd := dataStart + capacity*elemSize
	for i := dataStart; i < dataEnd; i++ {
		memory.Data()[i] = 0
	}

	return &Array[T]{
		memory:   memory,
		name:     name,
		offset:   offset,
		capacity: capacity,
		elemSize: elemSize,
	}, nil
}

// OpenArray opens an existing array in shared memory.
func OpenArray[T Numeric](memory *Memory, name string) (*Array[T], error) {
	entry := memory.Find(name)
	if entry == nil {
		return nil, fmt.Errorf("array '%s' not found", name)
	}

	var zero T
	elemSize := int(unsafe.Sizeof(zero))

	// Read capacity from header
	offset := int(entry.Offset)
	capacity := binary.LittleEndian.Uint64(memory.Data()[offset : offset+8])

	// Validate size
	expectedSize := ArrayHeaderSize + int(capacity)*elemSize
	if int(entry.Size) != expectedSize {
		return nil, fmt.Errorf("size mismatch: expected %d, got %d (wrong element type?)",
			expectedSize, entry.Size)
	}

	return &Array[T]{
		memory:   memory,
		name:     name,
		offset:   offset,
		capacity: int(capacity),
		elemSize: elemSize,
	}, nil
}

// Get returns the element at the given index.
func (a *Array[T]) Get(index int) T {
	if index < 0 || index >= a.capacity {
		panic(fmt.Sprintf("index %d out of bounds (capacity=%d)", index, a.capacity))
	}

	dataOffset := a.offset + ArrayHeaderSize + index*a.elemSize
	return *(*T)(unsafe.Pointer(&a.memory.Data()[dataOffset]))
}

// Set sets the element at the given index.
func (a *Array[T]) Set(index int, value T) {
	if index < 0 || index >= a.capacity {
		panic(fmt.Sprintf("index %d out of bounds (capacity=%d)", index, a.capacity))
	}

	dataOffset := a.offset + ArrayHeaderSize + index*a.elemSize
	*(*T)(unsafe.Pointer(&a.memory.Data()[dataOffset])) = value
}

// Capacity returns the array capacity.
func (a *Array[T]) Capacity() int {
	return a.capacity
}

// Name returns the array name.
func (a *Array[T]) Name() string {
	return a.name
}

// Fill sets all elements to the given value.
func (a *Array[T]) Fill(value T) {
	for i := 0; i < a.capacity; i++ {
		a.Set(i, value)
	}
}

// Slice returns a copy of the array data as a Go slice.
func (a *Array[T]) Slice() []T {
	result := make([]T, a.capacity)
	for i := 0; i < a.capacity; i++ {
		result[i] = a.Get(i)
	}
	return result
}

// CopyFrom copies data from a Go slice into the array.
func (a *Array[T]) CopyFrom(data []T) {
	n := len(data)
	if n > a.capacity {
		n = a.capacity
	}
	for i := 0; i < n; i++ {
		a.Set(i, data[i])
	}
}

// Range iterates over elements and calls fn for each.
func (a *Array[T]) Range(fn func(index int, value T) bool) {
	for i := 0; i < a.capacity; i++ {
		if !fn(i, a.Get(i)) {
			break
		}
	}
}
